strange behavior with similar keys

not sure if this is a bug in sophia itself, or in sphia(1), but imo, this behavior is very wrong:

``` c

#include <stdlib.h>
#include <assert.h>
#include "../src/sphia.h"

int
main(int argc, char *argv[]) {
  sphia_t * sphia = sphia_new("./test-db");
  assert(sphia);

  char k1[] = "key-1";
  char v1[] = "val-1";
  assert(-1 != sphia_set(sphia, k1, v1));

  char k2[] = "key-10";
  char v2[] = "val-10";
  assert(-1 != sphia_set(sphia, k2, v2));

  char *res = sphia_get(sphia, k1);
  printf("%s == %s\n", v1, res);
  assert(0 == strcmp(v1, res));

  sphia_free(sphia);
  return 0;
}
```

any ideas?

btw:

``` sh

./test/set-similar-keys
val-1 == val-10
Assertion failed: (0 == strcmp(v1, res)), function main, file test/set-similar-keys.c, line 22.
make: *** [test/set-similar-keys.c] Abort trap: 6

```

--- a/src/set.c
+++ b/src/set.c
@@ -18,5 +18,5 @@ sphia_set (sphia_t *sphia, char *key, char *value) {
   key[strlen(key)] = '\0';
   value[strlen(value)] = '\0';
 
-  return sp_set(sphia->db, key, strlen(key), value, strlen(value));
+  return sp_set(sphia->db, key, strlen(key) + 1, value, strlen(value) + 1);
 }

--- a/src/set.c
+++ b/src/set.c
@@ -15,8 +15,5 @@ sphia_set (sphia_t *sphia, char *key, char *value) {
   key = trim(key);
   value = trim(value);
 
-  key[strlen(key)] = '\0';
-  value[strlen(value)] = '\0';
-
   return sp_set(sphia->db, key, strlen(key) + 1, value, strlen(value) + 1);
 }

--- a/test/set-non-terminated-string.c
+++ b/test/set-non-terminated-string.c
@@ -1,0 +1,21 @@
+
+#include <stdlib.h>
+#include "sphia-test.h"
+
+static void
+test_non_terminated_string() {
+  char key[3] = { 'k', 'e', 'y' };
+  char value[] = "value";
+  assert(-1 != sphia_set(sphia, key, value));
+  char *res = sphia_get(sphia, key);
+  assert(0 == strcmp(value, res));
+}
+
+
+int
+main(int argc, char *argv[]) {
+  test_setup();
+  test_non_terminated_string();
+  test_teardown();
+  return 0;
+}

